odoo/addons/base/models/ir_http.py

```diff
@@ -325,6 +325,12 @@ def _pre_dispatch(cls, rule, args):
 
     @classmethod
     def _dispatch(cls, endpoint):
+        # Verify the captcha in case it was set on @http.route
+        # https://httpwg.org/specs/rfc9110.html#safe.methods
+        SAFE_HTTP_METHODS = ('GET', 'HEAD', 'OPTION', 'TRACE')
+        captcha = endpoint.routing.get('captcha')
+        if captcha and request.httprequest.method not in SAFE_HTTP_METHODS:
+            request.env['ir.http']._verify_request_recaptcha_token(captcha)
```

Comment:
He is right about http.py HttpDispatcher dispatch, that code only runs with `type='http'` endpoints.

Reply:
Please refer to the (enormous) docstring of http.py, copy/paste here for ease

```
Here be dragons:

    Application.__call__
        if path is like '/<module>/static/<path>':
            ...
        elif not request.db:
            ...
        else:
            Request._serve_db
                env['ir.http']._match
                if not match:
                    ...
                else:
                    Request._transactioning
                        model.retrying
                            env['ir.http']._authenticate
                            env['ir.http']._pre_dispatch
                            Dispatcher.pre_dispatch
                            Dispatcher.dispatch
                                env['ir.http']._dispatch        # <----- here
                                    route_wrapper
                                        endpoint
                            env['ir.http']._post_dispatch
```
```
ir.http._dispatch/Dispatcher.dispatch
  Deserialize the HTTP request body into ``request.params`` according to
  @route(type=...), call the controller endpoint, serialize its return
  value into an HTTP Response object.
```

Reply:
> I recall why it is inside _dispatch!! :D
So (1) it uses the database (ir.config_parameter) so it must be inside ir.http and not http.py, (2) it uses request.params so it must be after we set params which is done by http.py Dispatcher.dispatch. It leaves a single place: ir.http._dispatch.

Well, the csrf token validation also uses `ir.config_parameter` and it is inside `http.py`, it uses
`self.request.validate_csrf`, which uses the db with `self.env['ir.config_parameter'].sudo().get_param('database.secret')`

So the fact it uses the database/config parameter doesn't seem a valid reason to say that it **must** be inside `ir.http` rather than `http.py`.
I am not saying that this is the wrong place, don't get me wrong.
But to me the validation of the CSRF token and the captcha token should be at the same place for uniformity.
It's kinda the same thing.

So maybe it should be the csrf token validation which should be moved to `ir.http`.
That being said, as Bram invoked, this place is for both HTTP and JSON, while the CSRF is only for HTTP, and it would mean that if we move csrf here, it would require an `if http`, while where it is currently it doesn't need it. So for strict performance reason it could be left where it is then 🤷.